File: web-ui/lib/api/knowledge-graph.ts
import { apiClient } from "@/lib/api-client";

export interface ChangeSetItem {
  entity_type: string;
  entity_id?: number | null;
  action?: "added" | "updated" | "removed" | "schema_change";
  change_scope?: "local" | "structural" | "schema";
  details?: Record<string, unknown>;
}

export interface KnowledgeGraphBuild {
  id: number;
  graph_id: number;
  tenant_id: number;
  project_id: number;
  trigger_type: string;
  build_type: string;
  requested_by: number | null;
  status: string;
  stage: string;
  progress: number;
  error_code: string | null;
  safe_error_message: string | null;
  retry_attempt: number;
  worker_id: string | null;
  candidate_version_id: number | null;
  source_checkpoint: Record<string, unknown> | null;
  affected_entity_summary: Record<string, unknown> | null;
  queued_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeGraphVersion {
  id: number;
  graph_id: number;
  tenant_id: number;
  project_id: number;
  version_number: number;
  build_id: number | null;
  status: string;
  build_type: string;
  source_fingerprint: string | null;
  node_count: number;
  edge_count: number;
  disconnected_component_count: number;
  storage_reference: string | null;
  created_by: number | null;
  activated_at: string | null;
  superseded_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeGraphHealthCheck {
  id: number;
  graph_id: number | null;
  version_id: number | null;
  tenant_id: number;
  project_id: number;
  status: string;
  check_type: string;
  node_count: number;
  edge_count: number;
  orphan_ratio: number | null;
  disconnected_components: number;
  structural_checks: Record<string, unknown> | null;
  source_alignment: Record<string, unknown> | null;
  dependency_checks: Record<string, unknown> | null;
  warnings: string[] | null;
  errors: string[] | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface KnowledgeGraphStatus {
  project_id: number;
  graph_id: number | null;
  lifecycle_status: string;
  enabled: boolean;
  active_version_id: number | null;
  active_version_number: number | null;
  last_healthy_version_id: number | null;
  last_healthy_version_number: number | null;
  current_source_fingerprint: string | null;
  active_source_fingerprint: string | null;
  last_successful_build_at: string | null;
  last_health_check_at: string | null;
  active_node_count: number;
  active_edge_count: number;
  health_status: string;
  has_active_version: boolean;
  builds: KnowledgeGraphBuild[];
  versions: KnowledgeGraphVersion[];
}

export interface KnowledgeGraphRebuildResponse {
  build: KnowledgeGraphBuild;
  build_type: string;
  enqueued: boolean;
}

export interface ExecutiveInsightDependency {
  ready: boolean;
  mode: "full" | "limited" | "blocked";
  graph_status: string;
  graph_version_id: number | null;
  graph_version_number: number | null;
  active_node_count: number;
  active_edge_count: number;
  warnings: string[];
  blocking_reasons: string[];
  disclosure: string;
  health_status: string;
}

export const knowledgeGraphApi = {
  status: (projectId: string | number): Promise<KnowledgeGraphStatus> =>
    apiClient.get<KnowledgeGraphStatus>(`/api/projects/${projectId}/knowledge-graph/status`),

  rebuild: (projectId: string | number): Promise<KnowledgeGraphRebuildResponse> =>
    apiClient.post<KnowledgeGraphRebuildResponse>(`/api/projects/${projectId}/knowledge-graph/rebuild`, {}),

  rebuildIncremental: (
    projectId: string | number,
    changeSet: ChangeSetItem[],
    reason?: string,
  ): Promise<KnowledgeGraphRebuildResponse> =>
    apiClient.post<KnowledgeGraphRebuildResponse>(
      `/api/projects/${projectId}/knowledge-graph/rebuild/incremental`,
      { change_set: changeSet, reason },
    ),

  listBuilds: (projectId: string | number): Promise<KnowledgeGraphBuild[]> =>
    apiClient.get<KnowledgeGraphBuild[]>(`/api/projects/${projectId}/knowledge-graph/builds`),

  getBuild: (projectId: string | number, buildId: number): Promise<KnowledgeGraphBuild> =>
    apiClient.get<KnowledgeGraphBuild>(`/api/projects/${projectId}/knowledge-graph/builds/${buildId}`),

  runHealthCheck: (projectId: string | number): Promise<KnowledgeGraphHealthCheck> =>
    apiClient.post<KnowledgeGraphHealthCheck>(`/api/projects/${projectId}/knowledge-graph/health-check`, {}),

  getHealth: (projectId: string | number): Promise<KnowledgeGraphHealthCheck> =>
    apiClient.get<KnowledgeGraphHealthCheck>(`/api/projects/${projectId}/knowledge-graph/health`),

  listVersions: (projectId: string | number): Promise<KnowledgeGraphVersion[]> =>
    apiClient.get<KnowledgeGraphVersion[]>(`/api/projects/${projectId}/knowledge-graph/versions`),

  executiveInsightDependency: (
    projectId: string | number,
  ): Promise<ExecutiveInsightDependency> =>
    apiClient.get<ExecutiveInsightDependency>(
      `/api/projects/${projectId}/knowledge-graph/dependencies/executive-insight`,
    ),
};

File: platform-api/app/schemas/knowledge_graph.py
"""Pydantic schemas for the knowledge graph lifecycle API."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class ChangeSetItem(BaseModel):
    entity_type: str
    entity_id: int | None = None
    action: str = "updated"  # added | updated | removed | schema_change
    change_scope: str = "local"  # local | structural | schema
    details: dict[str, Any] | None = None


class IncrementalRebuildRequest(BaseModel):
    change_set: list[ChangeSetItem] = Field(default_factory=list)
    reason: str | None = None


class KnowledgeGraphBuildRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    graph_id: int
    tenant_id: int
    project_id: int
    trigger_type: str
    build_type: str
    requested_by: int | None
    status: str
    stage: str
    progress: int
    error_code: str | None
    safe_error_message: str | None
    retry_attempt: int
    worker_id: str | None
    candidate_version_id: int | None
    source_checkpoint: dict[str, Any] | None
    affected_entity_summary: dict[str, Any] | None
    # KG-48: per-stage duration breakdown (ms) + source counts, so an
    # operator can see which stage was slow or where a build failed without
    # reading raw logs.
    stage_metrics: dict[str, Any] | None = None
    queued_at: datetime | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class KnowledgeGraphVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    graph_id: int
    tenant_id: int
    project_id: int
    version_number: int
    build_id: int | None
    status: str
    build_type: str
    source_fingerprint: str | None
    node_count: int
    edge_count: int
    disconnected_component_count: int
    storage_reference: str | None
    created_by: int | None
    activated_at: datetime | None
    superseded_at: datetime | None
    created_at: datetime
    updated_at: datetime
    # KG-11/KG-15: validation errors/warnings plus the per-source-type
    # coverage manifest (total/included/excluded/failed/pending), so a
    # "succeeded" build can never silently conceal truncated, failed, or
    # still-processing sources.
    validation_summary: dict[str, Any] | None = None


class KnowledgeGraphHealthCheckRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    graph_id: int | None
    version_id: int | None
    tenant_id: int
    project_id: int
    status: str
    check_type: str
    node_count: int
    edge_count: int
    orphan_ratio: float | None
    disconnected_components: int
    structural_checks: dict[str, Any] | None
    source_alignment: dict[str, Any] | None
    dependency_checks: dict[str, Any] | None
    source_coverage: dict[str, Any] | None
    warnings: list[str] | None
    errors: list[str] | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime


class KnowledgeGraphStatusRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    project_id: int
    graph_id: int | None
    lifecycle_status: str
    enabled: bool
    active_version_id: int | None
    active_version_number: int | None
    last_healthy_version_id: int | None
    last_healthy_version_number: int | None
    current_source_fingerprint: str | None
    active_source_fingerprint: str | None
    last_successful_build_at: datetime | None
    last_health_check_at: datetime | None
    active_node_count: int
    active_edge_count: int
    health_status: str
    has_active_version: bool
    builds: list[KnowledgeGraphBuildRead]
    versions: list[KnowledgeGraphVersionRead]


class KnowledgeGraphRebuildResponse(BaseModel):
    build: KnowledgeGraphBuildRead
    build_type: str
    enqueued: bool


class ExecutiveInsightDependencyRead(BaseModel):
    ready: bool
    mode: str  # full | limited | blocked
    graph_status: str
    graph_version_id: int | None
    graph_version_number: int | None
    active_node_count: int
    active_edge_count: int
    warnings: list[str]
    blocking_reasons: list[str]
    disclosure: str
